### ISSUE TITLE
Find bug when using SwapLineUpCommand & SwapLineDownCommand

### ISSUE THREAD

**Issue by brupelo** [brupelo/pyblime#23]
Consider this mcve:

    import sys
    from pathlib import Path

    from PyQt5.Qt import *

    from pyblime import View, center, data_path
    from pysyntect import SyntaxDefinition, SyntaxSetBuilder

    import sublime
    import sublime_plugin

    def expand_to_line(view, region):
        """
        As view.full_line, but doesn't expand to the next line if a full line is
        already selected
        """

        if not (region.a == region.b) and view.substr(region.end() - 1) == '\n':
            return sublime.Region(view.line(region).begin(), region.end())
        else:
            return view.full_line(region)

    def extract_line_blocks(view):
        blocks = [expand_to_line(view, s) for s in view.sel()]
        if len(blocks) == 0:
            return blocks

        # merge any adjacent blocks
        merged_blocks = [blocks[0]]
        for block in blocks[1:]:
            last_block = merged_blocks[-1]
            if block.begin() <= last_block.end():
                merged_blocks[-1] = sublime.Region(last_block.begin(), block.end())
            else:
                merged_blocks.append(block)

        return merged_blocks

    class SwapLineUpCommand(sublime_plugin.TextCommand):
        def run(self, edit):
            blocks = extract_line_blocks(self.view)

            # No selection
            if len(blocks) == 0:
                return

            # Already at BOF
            if blocks[0].begin() == 0:
                return

            # Add a trailing newline if required, the logic is simpler if every line
            # ends with a newline
            add_trailing_newline = (self.view.substr(self.view.size() - 1) != '\n') and blocks[-1].b == self.view.size()
            if add_trailing_newline:
                # The insert can cause the selection to move. This isn't wanted, so
                # reset the selection if it has moved to EOF
                sel = [r for r in self.view.sel()]
                self.view.insert(edit, self.view.size(), '\n')
                if self.view.sel()[-1].end() == self.view.size():
                    # Selection has moved, restore the previous selection
                    self.view.sel().clear()
                    for r in sel:
                        self.view.sel().add(r)

                # Fix up any block that should now include this newline
                blocks[-1] = sublime.Region(blocks[-1].a, blocks[-1].b + 1)

            # Process in reverse order
            blocks.reverse()
            for b in blocks:
                prev_line = self.view.full_line(b.begin() - 1)
                self.view.insert(edit, b.end(), self.view.substr(prev_line))
                self.view.erase(edit, prev_line)

            if add_trailing_newline:
                # Remove the added newline
                self.view.erase(edit, sublime.Region(self.view.size() - 1, self.view.size()))

            # Ensure the selection is visible
            self.view.show(self.view.sel(), False)

    class SwapLineDownCommand(sublime_plugin.TextCommand):
        def run(self, edit):
            blocks = extract_line_blocks(self.view)

            # No selection
            if len(blocks) == 0:
                return

            # Already at EOF
            if blocks[-1].end() == self.view.size():
                return

            # Add a trailing newline if required, the logic is simpler if every line
            # ends with a newline
            add_trailing_newline = (self.view.substr(self.view.size() - 1) != '\n')
            if add_trailing_newline:
                # No block can be at EOF (checked above), so no need to fix up the
                # blocks
                self.view.insert(edit, self.view.size(), '\n')

            # Process in reverse order
            blocks.reverse()
            for b in blocks:
                next_line = self.view.full_line(b.end())
                contents = self.view.substr(next_line)

                self.view.erase(edit, next_line)
                self.view.insert(edit, b.begin(), contents)

            if add_trailing_newline:
                # Remove the added newline
                self.view.erase(edit, sublime.Region(self.view.size() - 1, self.view.size()))

            # Ensure the selection is visible
            self.view.show(self.view.sel(), False)

    class Test(View):

        def __init__(self):
            super().__init__()
            builder = SyntaxSetBuilder()
            path = data_path() / "st_build_3149/syntax/Python.sublime-syntax"
            syntax = SyntaxDefinition.load_from_str(path.read_text(), True)
            builder.add(syntax)
            syntax_set = builder.build()
            self.load_syntax_set(syntax_set)
            self["color_scheme"] = data_path() / "st_build_3149/themes/Monokai.tmTheme"
            self.setWindowTitle(Path(__file__).stem)
            self.load_file(__file__)

    if __name__ == "__main__":
        app = QApplication(sys.argv)

        view = Test()
        view.register_command("alt+up", SwapLineUpCommand)
        view.register_command("alt+down", SwapLineDownCommand)
        view.resize(800, 768)

        center(view)
        view.show_()
        sys.exit(app.exec_())

Right now is needed to find the bug that makes `SwapLineUpCommand` and `SwapLineDownCommand` behaving identically to Sublime, here's a showcase showing the differences:

![showcase](https://dl.dropboxusercontent.com/s/fcvbh5irxsbxmrc/2019-03-30_14-37-22.gif)

once the bug is found we'd need to add the proper cases to the unit tests until those commands behave exactly the same than Sublime's

@keith-hall Would you be interested to get assigned to this one? This would be a nice good first issue... and to be honest? A pretty amazing one as it'd prove pyblime is starting to get pretty usable... as you can see those Sublime's command are not small tiny anymore ;)

Please let me know if you're interested so I'll move to dispatch another tickets myself... :)

**Comment by keith-hall** [comment]
sure, you can assign me :) not sure when I will work on it though

**Comment by brupelo** [comment]
Awesome!

Thanks, and sure, no rush at all... it's a matter of being synced as a team. That way I can focus on other tickets myself and I know on which stuff is "everybody" working on ;)

Plus, I hope thanks to this ticket you'll get more engaged to the project gradually, otherwise working all by myself is becoming boring ;)

Any question about the code or whatever, please don't doubt to ask... If you have never worked with pyqt it could take you time to become comfortable, so...

Anyway, let's see how it goes :D

Again, thanks for that pro-active attitude ;)

**Comment by brupelo** [comment]
sorry, I'm trying to figure out how to assign it to you but I'm still trying to figure out, do you know why I don't see your name here?

![showcase](https://dl.dropboxusercontent.com/s/hy51n5gywccuj5g/2019-03-30_18-33-58.gif)

**Comment by brupelo** [comment]
so it seems you need to become a collaborator to be able to get assigned then? Mmm, tricky :)

**Comment by keith-hall** [comment]
yeah its interesting... I guess its GitHub's way to make you pay to get features like locking collaborators from pushing directly to `master` etc. I guess I should still work on my fork and submit PRs when features are ready to be reviewed.

**Comment by brupelo** [comment]
Oh, thanks, that would make it easier for me, remember I'm not a github expert... and yeah, definitely using PRs will make the job as maintainer much much easier to me... in any case, if you feel pushing stuff directly to you will be faster don't hesitate, now you've got push rights yourself ;)
